src/qibo/transpiler/decompositions.py

```diff
@@ -20,43 +17,63 @@ def add(self, gate, decomposition):
         """Register a decomposition for a gate."""
         self.decompositions[gate] = decomposition
 
-    def count_2q(self, gate):
+    def count_2q(self, gate, backend):
         """Count the number of two-qubit gates in the decomposition of the given gate."""
         if gate.parameters:
-            decomposition = self.decompositions[gate.__class__](gate)
+            if isinstance(
+                gate,
+                (gates.FusedGate, gates.Unitary, gates.GeneralizedfSim, gates.fSim),
```

Comment:
Why do we need to treat these differently?

Reply:
We need the backend to implement the decomposition

Reply:
Yes that I can see myself, but why is that the case? It was not like this before. In general, I don't understand why you need the backend to count how many two qubit gates you have in the decomposition.

Reply:
It is not to count 2 qubit gates, it is to decompose the matrix (using u3_decomposition or two_qubit_decomposition) when calling the decomposition.

Reply:
Well but the purpose of the function is just to count the 2q gates as the name and the docstring says. Thus you don't even need to instantiate the gate in principle, and even if you do it does not really matter whether the parameters of the gate are `torch.Tensor`s or plain floats, but this is probably more related to the decomposition mechanism in general not specifically to this PR.

Anyway, at least we could try to wrap this snippet:
``` python
            if isinstance(
                gate,
                (gates.FusedGate, gates.Unitary, gates.GeneralizedfSim, gates.fSim),
            ):
                decomposition = self.decompositions[gate.__class__](gate, backend)
            else:
                decomposition = self.decompositions[gate.__class__](gate)
```
in a `_helper_function` since it is repeated several times.

Reply:
Not all decomposition have a fixed number of gates. When decomposing a two qubit unitary the number of 2q gates is usually 3 but in some cases it reduces to 2 or even 1. It is necessary to perform the decomposition to know it.

Reply:
Yes, of course, but the decomposition is well defined even though the gates that constitute it are not explicitely instantiated nor whether their parameters are consistent with the backend. The decomposition should be backend independent I believe.

Reply:
Making the decomposition backend independent is not part of this PR. However, I think that the decomposition should be implemented on the selected backend. This would make the gradients pass (in principle) and help in executing qml algorithms directly on hardware. Regarding the function count_2q/1q i agree that the backend is not fundamental, if you prefer i can set backend=NumpyBackend() at the beginning of the function and avoid passing it.